File: base/tests/test_hierarchisation.py
# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at
# sgpp.sparsegrids.org

import unittest
import pysgpp

#-------------------------------------------------------------------------------
# # tests the correctness of the hierarchisation and dehierachisation
# @param node1 the vector of the node base values before hierarchisation and dehierarchisation
# @param node2 the vector of the node base values after hierarchisation and dehierarchisation
# @return maximum error during the transformations
def testHierarchisationResults(node1, node2):
  error = 0.0

  for i in xrange(len(node1)):
    if abs(node1[i] - node2[i]) > error:
      error = abs(node1[i] - node2[i])

  return error


#-------------------------------------------------------------------------------
# # Hierarchise and dechierarchise a regular sparse grid for a given function and test.
# @param obj reference to unittest
# @param grid the grid object
# @param level the number of levels used in the grid
# @param function string of function which to use for test
def testHierarchisationDehierarchisation(obj, grid, level, function,
                                         evalOp=None):
  places = 7 if pysgpp.cvar.USING_DOUBLE_PRECISION else 4

  # generate a regular test grid
  generator = grid.createGridGenerator()
  generator.regular(level)

  storage = grid.getStorage()
  dim = storage.dim()

  # generate the node_values vector
  node_values = pysgpp.DataVector(storage.size())
  for n in xrange(storage.size()):
    points = storage.get(n).getCoordsString().split()
    node_values[n] = evalFunction(function, points)

  # do hierarchisation
  alpha = doHierarchisation(node_values, grid)

  # test hierarchisation
  p = pysgpp.DataVector(storage.dim())
  if evalOp is None:
    evalOp = pysgpp.createOperationEval(grid)
  for n in xrange(storage.size()):
    storage.get(n).getCoords(p)
    obj.assertAlmostEqual(evalOp.eval(alpha, p), node_values[n],
                places=places)

  # do dehierarchisation
  node_values_back = doDehierarchisation(alpha, grid)

  # test dehierarchisation
  obj.assertAlmostEqual(testHierarchisationResults(node_values, node_values_back),
              0.0, places=places)

#-------------------------------------------------------------------------------
# # Hierarchise and dechierarchise a regular sparse grid for a given function and test
# Difference from the function above is the getCoordsStretching function usage.
# @param obj reference to unittest
# @param grid the grid object
# @param level the number of levels used in the grid
# @param function string of function which to use for test
def testHierarchisationDehierarchisationStretching(obj, grid, level, function):
  places = 7 if pysgpp.cvar.USING_DOUBLE_PRECISION else 4

  # generate a regular test grid
  generator = grid.createGridGenerator()
  generator.regular(level)

  storage = grid.getStorage()
  dim = storage.dim()
  stretch = storage.getStretching()
  # generate the node_values vector
  node_values = pysgpp.DataVector(storage.size())
  for n in xrange(storage.size()):
    points = storage.get(n).getCoordsStringStretching(stretch).split()
    node_values[n] = evalFunction(function, points)

  # do hierarchisation
  alpha = doHierarchisation(node_values, grid)

  # test hierarchisation
  p = pysgpp.DataVector(storage.dim())
  evalOp = pysgpp.createOperationEval(grid)
  for n in xrange(storage.size()):
    storage.get(n).getCoordsStretching(p, stretch)
    obj.assertAlmostEqual(evalOp.eval(alpha, p), node_values[n],
                places=places)

  # do dehierarchisation
  node_values_back = doDehierarchisation(alpha, grid)

  # test dehierarchisation
  obj.assertAlmostEqual(testHierarchisationResults(node_values, node_values_back),
              0.0, places=places)

#-------------------------------------------------------------------------------
# # hierarchisation of the node base values on a grid
# @param node_values DataVector that holds the coefficients of the function's node base
# @param grid the grid matching to the node_vector
def doHierarchisation(node_values, grid):
  alpha = pysgpp.DataVector(node_values)
  hierarchisation = pysgpp.createOperationHierarchisation(grid)
  hierarchisation.doHierarchisation(alpha)
  return alpha


#-------------------------------------------------------------------------------
# # hierarchisation of the node base values on a grid
# @param alpha DataVector that holds the coefficients of the sparse grid's ansatzfunctions
# @param grid thee grid matching to the alpha vector
def doDehierarchisation(alpha, grid):
  node_values = pysgpp.DataVector(alpha)
  hierarchisation = pysgpp.createOperationHierarchisation(grid)
  hierarchisation.doDehierarchisation(node_values)
  return node_values


# # evalutes a given function
# @param function a string the gives the function; x1...xn must be the names of the placeholders
# @param points sorted list of the coordinates (x1...xn) of evaluation point
# @return returns the function value at points
def evalFunction(function, points):
  for i in xrange(len(points) - 1, -1, -1):
    function = function.replace("x{}".format(i + 1), points[i])

  return eval(function)


# # build parabola test function over [0,1]^d
# @param dim dimension of the parabola's space
# @return returns a string that contains the function as string
def buildParabola(dim):
  function = str(pow(4.0, dim))

  for i in xrange(dim):
    function += "*x{0}*(1-x{0})".format(i + 1)

  return function


# # build parabola test function over [0,1]^d with boundaries
# @param dim dimension of the parabola's space
# @return returns a string that contains the function as string
def buildParabolaBoundary(dim):
  function = "1.0"

  for i in xrange(dim):
    function += "*((0.25*(x{0}-0.7)*(x{0}-0.7))+2.0)".format(i + 1)

  return function


class TestHierarchisation(unittest.TestCase):
  def testHierarchisationLinear(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabola(dim)
      grid = pysgpp.Grid.createLinearGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationModLinear(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabola(dim)
      grid = pysgpp.Grid.createModLinearGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationModLinearWithBoundary(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabolaBoundary(dim)
      grid = pysgpp.Grid.createModLinearGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationTruncatedBoundary(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabolaBoundary(dim)
      grid = pysgpp.Grid.createLinearTruncatedBoundaryGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationBoundary(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabolaBoundary(dim)
      grid = pysgpp.Grid.createLinearBoundaryGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationStretchedTruncatedBoundary1D(self):
    dim = 1
    level = 5
    function = buildParabolaBoundary(dim)
    str1d = pysgpp.Stretching1D()
    str1d.type = 'log'
    str1d.x_0 = 0
    str1d.xsi = 10
    dimBound = pysgpp.DimensionBoundary()
    dimBound.leftBoundary = 0.00001
    dimBound.rightBoundary = 1
    stretch = pysgpp.Stretching(1, dimBound, str1d)

    grid = pysgpp.Grid.createLinearStretchedTruncatedBoundaryGrid(1)
    grid.getStorage().setStretching(stretch)

    testHierarchisationDehierarchisationStretching(self, grid, level, function)

  def testHierarchisationStretchedTruncatedBoundaryD(self):
    str1d = pysgpp.Stretching1D()
    str1d.type = 'sinh'
    str1d.x_0 = 1
    str1d.xsi = 10
    dimBound = pysgpp.DimensionBoundary()
    dimBound.leftBoundary = 0.001
    dimBound.rightBoundary = 1

    dim = 3
    level = 5

    dimBoundVector = pysgpp.DimensionBoundaryVector(3)
    dimBoundVector[0] = dimBound
    dimBoundVector[1] = dimBound
    dimBoundVector[2] = dimBound

    str1dVector = pysgpp.Stretching1DVector(3)
    str1dVector[0] = str1d
    str1dVector[1] = str1d
    str1dVector[2] = str1d

    stretch = pysgpp.Stretching(dim, dimBoundVector, str1dVector)

    function = buildParabolaBoundary(dim)
    grid = pysgpp.Grid.createLinearStretchedTruncatedBoundaryGrid(dim)
    grid.getStorage().setStretching(stretch)

    testHierarchisationDehierarchisationStretching(self, grid, level, function)

  def testHierarchisationPrewavelet(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabola(dim)
      grid = pysgpp.Grid.createPrewaveletGrid(dim)
      testHierarchisationDehierarchisation(self, grid, level, function)

  def testHierarchisationFundamentalSpline(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabola(dim)

      for degree in [1, 3, 5]:
        grid = pysgpp.Grid.createFundamentalSplineGrid(dim, degree)
        evalOp = pysgpp.createOperationNaiveEval(grid)
        testHierarchisationDehierarchisation(self, grid, level, function,
                                             evalOp)

  def testHierarchisationModFundamentalSpline(self):
    level = 5

    for dim in [1, 3]:
      function = buildParabola(dim)

      for degree in [1, 3, 5]:
        grid = pysgpp.Grid.createModFundamentalSplineGrid(dim, degree)
        evalOp = pysgpp.createOperationNaiveEval(grid)
        testHierarchisationDehierarchisation(self, grid, level, function,
                                             evalOp)

# Run tests for this file if executed as application
if __name__ == '__main__':
  unittest.main()

File: base/tests/test_hierarchisation.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#define BOOST_TEST_DYN_LINK

#include <boost/test/unit_test.hpp>

#include <sgpp/base/datatypes/DataVector.hpp>
#include <sgpp/base/grid/Grid.hpp>
#include <sgpp/base/operation/BaseOpFactory.hpp>

#include <vector>

using sgpp::base::DataVector;
using sgpp::base::DimensionBoundary;
using sgpp::base::Grid;
using sgpp::base::GridGenerator;
using sgpp::base::GridStorage;
using sgpp::base::OperationEval;
using sgpp::base::OperationHierarchisation;
using sgpp::base::OperationNaiveEval;
using sgpp::base::Stretching;
using sgpp::base::Stretching1D;

void testHierarchisationDehierarchisation(sgpp::base::Grid& grid, size_t level,
                                          double (*func)(DataVector&), double tolerance = 0.0,
                                          bool naiveOp = false, bool doStretch = false) {
  grid.getGenerator().regular(level);
  GridStorage& gridStore = grid.getStorage();
  size_t dim = gridStore.getDimension();
  Stretching* stretch = NULL;

  if (doStretch) stretch = gridStore.getStretching();

  DataVector node_values = DataVector(gridStore.getSize());
  DataVector coords = DataVector(dim);

  for (size_t n = 0; n < gridStore.getSize(); n++) {
    if (doStretch) {
      gridStore.get(n)->getCoordsStretching(coords, *stretch);
    } else {
      gridStore.get(n)->getCoords(coords);
    }

    node_values[n] = func(coords);
  }

  DataVector alpha = DataVector(node_values);
  std::unique_ptr<OperationHierarchisation> hierarchisation(
      sgpp::op_factory::createOperationHierarchisation(grid));
  hierarchisation->doHierarchisation(alpha);

  if (naiveOp == true) {
    std::unique_ptr<OperationNaiveEval> op(sgpp::op_factory::createOperationNaiveEval(grid));

    for (size_t n = 0; n < gridStore.getSize(); n++) {
      if (doStretch) {
        gridStore.get(n)->getCoordsStretching(coords, *stretch);
      } else {
        gridStore.get(n)->getCoords(coords);
      }

      double eval = op->eval(alpha, coords);
      BOOST_CHECK_CLOSE(eval, node_values[n], tolerance);
    }
  } else {
    std::unique_ptr<OperationEval> op(sgpp::op_factory::createOperationEval(grid));

    for (size_t n = 0; n < gridStore.getSize(); n++) {
      if (doStretch) {
        gridStore.get(n)->getCoordsStretching(coords, *stretch);
      } else {
        gridStore.get(n)->getCoords(coords);
      }

      double eval = op->eval(alpha, coords);
      BOOST_CHECK_CLOSE(eval, node_values[n], tolerance);
    }
  }

  DataVector node_values_back = DataVector(alpha);
  hierarchisation->doDehierarchisation(node_values_back);

  for (size_t n = 0; n < gridStore.getSize(); n++) {
    BOOST_CHECK_CLOSE(node_values_back[n], node_values[n], tolerance);
  }
}

double parabola(DataVector& input) {
  double result = 1.;

  for (size_t i = 0; i < input.getSize(); i++) {
    result *= input[i] * (1. - input[i]) * 4.;
  }

  return result;
}

double parabolaBoundary(DataVector& input) {
  double result = 1.;

  for (size_t i = 0; i < input.getSize(); i++) {
    result *= 0.25 * (input[i] - 0.7) * (input[i] - 0.7) + 2;
  }

  return result;
}

BOOST_AUTO_TEST_SUITE(testHierarchization)

BOOST_AUTO_TEST_CASE(testHierarchisationLinear) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createLinearGrid(dim);
    testHierarchisationDehierarchisation(*grid, level, &parabola);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationModLinear) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createModLinearGrid(dim);
    testHierarchisationDehierarchisation(*grid, level, &parabola);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationModLinearWithBoundary) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createModLinearGrid(dim);
    testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-12, true);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationTruncatedBoundary) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createLinearBoundaryGrid(dim);
    testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-12, true);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationBoundary) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createLinearBoundaryGrid(dim, 0);
    testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-12);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationPrewavelet) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    std::unique_ptr<Grid> grid = Grid::createPrewaveletGrid(dim);
    testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-12);
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationFundamentalSpline) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    for (int degree = 1; degree < 6; degree += 2) {
      std::unique_ptr<Grid> grid = Grid::createFundamentalSplineGrid(dim, degree);
      testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-9, true);
    }
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationModFundamentalSpline) {
  int level = 5;

  for (int dim = 1; dim < 4; dim++) {
    for (int degree = 1; degree < 6; degree += 2) {
      std::unique_ptr<Grid> grid = Grid::createModFundamentalSplineGrid(dim, degree);
      testHierarchisationDehierarchisation(*grid, level, &parabola, 1e-9, true);
    }
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationPoly) {
  int level = 5;
  int degree[4] = {2, 3, 5, 8};

  for (int dim = 1; dim < 4; dim++) {
    for (int i = 0; i < 4; i++) {
      std::unique_ptr<Grid> grid = Grid::createPolyGrid(dim, degree[i]);
      testHierarchisationDehierarchisation(*grid, level, &parabola, 0.0, true);
    }
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationPolyTruncatedBoundary) {
  int level = 5;
  int degree[4] = {2, 3, 5, 8};

  for (int dim = 1; dim < 4; dim++) {
    for (int i = 0; i < 4; i++) {
      std::unique_ptr<Grid> grid = Grid::createPolyBoundaryGrid(dim, degree[i]);
      testHierarchisationDehierarchisation(*grid, level, &parabola, 0.0, true);
    }
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationModPoly) {
  int level = 5;
  int degree[4] = {2, 3, 5, 8};

  for (int dim = 1; dim < 4; dim++) {
    for (int i = 0; i < 4; i++) {
      std::unique_ptr<Grid> grid = Grid::createModPolyGrid(dim, degree[i]);
      testHierarchisationDehierarchisation(*grid, level, &parabola, 0.0, true);
    }
  }
}

BOOST_AUTO_TEST_CASE(testHierarchisationStretchedTruncatedBoundary1D) {
  int dim = 1;
  int level = 5;
  Stretching1D str1d;
  str1d.type = "log";
  str1d.x_0 = 0.;
  str1d.xsi = 10;
  DimensionBoundary dimBound;
  dimBound.leftBoundary = 0.00001;
  dimBound.rightBoundary = 1.;
  Stretching stretch(dim, &dimBound, &str1d);
  std::unique_ptr<Grid> grid = Grid::createLinearStretchedBoundaryGrid(dim);
  grid->getStorage().setStretching(stretch);
  testHierarchisationDehierarchisation(*grid, level, &parabolaBoundary, 1e-13, false, true);
}

BOOST_AUTO_TEST_CASE(testHierarchisationStretchedTruncatedBoundary3D) {
  int dim = 3;
  int level = 5;
  Stretching1D str1d;  // = new Stretching1D();
  str1d.type = "sinh";
  str1d.x_0 = 1.;
  str1d.xsi = 10;
  DimensionBoundary dimBound;  // = new DimensionBoundary();
  dimBound.leftBoundary = 0.001;
  dimBound.rightBoundary = 1.;

  std::vector<Stretching1D> stretch_vec;
  stretch_vec.push_back(str1d);
  stretch_vec.push_back(str1d);
  stretch_vec.push_back(str1d);
  std::vector<DimensionBoundary> dimBound_vec;
  dimBound_vec.push_back(dimBound);
  dimBound_vec.push_back(dimBound);
  dimBound_vec.push_back(dimBound);

  Stretching stretch(dim, dimBound_vec, stretch_vec);
  std::unique_ptr<Grid> grid = Grid::createLinearStretchedBoundaryGrid(dim);
  grid->getStorage().setStretching(stretch);
  testHierarchisationDehierarchisation(*grid, level, &parabolaBoundary, 1e-12, false, true);
}

BOOST_AUTO_TEST_SUITE_END()
